### DISCUSSION TITLE
req.cookies is not able to parse cookie when multiple cookies are present

### DISCUSSION THREAD

**Discussion by jagyas** [vercel/next.js#15084] [Help] [answered]
# Bug report

## Describe the bug

I tried to access cookie in next.js server side **api** by using **req.cookies** helper but when multiple cookies are present in request header **req.cookies** is giving all the cookies and all cookies are in encrypted form. 

I tried to print the **req.cookies** result in console log and below is the output:
`{"_gid":"GA1.2.595108422.1594311424","gitpod-user":"loggedIn","theme":"{\"id\":\"dark\",\"mode\":\"dark\",\"colors\":{\"brand\":\"\",\"brand2\":\"\",\"background1\":\"#1e1e1e\",\"background2\":\"\",\"background3\":\"\",\"paperShadow\":\"\",\"fontColor1\":\"\",\"fontColor2\":\"\",\"fontColor3\":\"\",\"disabled\":\"\"}}","a0:session":"Fe26.2**9014ce40bb1bfe378ecc9a239b6423f4673c9cbc2e118daefb296fabd3e5fe9f*UsQNbXQEDoMUpG4VuYNPtg*eZmsbdH2SIfc3_arJYk86fzKICsR0ZO_AwaYT450osQKxEqBfRwYEaSgVy8oFRl2ya3uth-7HuQYfsh-4yPdR3V5qeFvyRoYk3Q70zs6Ot0sz-6YoJWq2dAZtBaCHNDHZnOt4qrBqBI5eV9hjbm-kBBqpcqN7d6r9kd1vXZLbtok71OOtDE1MWWkFr2dWbkQS1Vi2rHR7zTX17sevVwtD7IyrlnsmrZr6wKm5x8D71wdOH5H66nxFjkufUx0VxUTUNu9ouGopB4Li722-6WzRwumNEqBINW7nqr8ubA3KYgGmECBghufcuBXrkNOlv0S8bNqn1cmimxRSGo-f3nyjeVdlsAek1vwUZ0EId97s_mjPz-2ZEsD0V5QyGB1bdUVDFmSugvbxMsBPCuIImv6l7dpeGh5OV_HEfQws3JV-hCTTGUPE9akbXA_akr4yhYiIodIVOQAqFxsJWYjB87pKF3GC6v1be6wtARSsRpUZ92snlfamTv874pi1Bh6iEK72z8o9BC7FDB7EoA3ZNkR9_CSdOMIt-ZgIEoqjXMKWwog2C72Pjmxp_6sFBazebMwVeTpqNIrKnkZusrTryukImP6VgTFEj6o1L9RJZqyLG_cxQ4sYlT6UB6sQe36qI3tWpDEUWQ9XPcxhO9nnjyBDj2lpD2_rFy7cS6F7EvaKPrKCRSYQnv4Av_G9q7NYXH9pAzA_-W8MWkOMLJcinsZ-MkqpIwDvGScHDLhnvbdg4QHJvHpAJ9h2Jl4oDveMjYVqKsCft0zhm5QtoUEb9SPV1w6GaBkuinzkaDoOCBOwYbAcL_Sz4UJbHl8KiqLCp1cuwQ7nsRPgfAuktpI-nChhSkk3U1Cl7ZS-tMPn885kKdedWnXnXaHV-odkRnF1fdt0NObcL8sVXC_LXxsysGDCHsCoo3SKCzKYJODoT8oW10KMSOHOWA9sMv3Nwj9CDX8-qQ1X4tu_0KRFTL33VhbrZMmize3omOoZBD0UeVcdYTvUDKa2gI-POyemw2Ck4btl2LJFqTsIte4lvUbFGP6tN-tXzy1JwQCZKlbMgNqVxcmPDuxtDiVQjklBQ6OkJ-R5ceJK--vpPuQJ1jTdKxU6H-LlQpiYuZTuU9D9lSaq3eb7H-2mWBJ648NneXILlofq8Mty0zS4fdhvVrmwk8C9ZMuu3fqdpZMErZprAEk_rMJkVjBqnYZ3OdWhjXP9ZNajvZDY-XsHNjNTnJx5AwL9p9WObauMS3oWdfKfDcYcPyNbgE_HXLxSBanqusEIXkdV_JyatCIdTLgNvrFdVuda0ol7smNEws7LSM4NrtMWYUZvsy4Hd6jHZk4X4_zw30O1DKaOt2RGsGWz1cVVrkDv6XVCqng7xW7VLL6qlGhDlQirS-f11wmdDCZ2-RKD65yl1_54_IgwxVawdTZaKt_B5r_F-Do-GEHj8Pujl9PZFWOIILyYMwAG6jjil87zlt-HV_d_yHq6DcK0YazahraQ5Ob7KRfV8XSwciB3tan0IJlaMybE68SAUHScN9KsEPCKCcC6XGqSQPCJPTla6ObiAu8yYc0hTXqLjGKXJ_XzVwle3AVsbDOBPUSpTuoCEuSzakdOyPnIdUFuNuaclKvOjT7gVEA9NoZsBJ7MybLrPlhHjENKFM9mWSg2jfUynelDDnoJ7ZTKSmIHZaiCmb4B-iq1RrzpQVh5KgzWya2lVquMGACwVnlRfH2lOxuYSyHrdrtf_oiEpJz-P3WxLrewZGhuom8ojU6xCwKqkJ128MaEz7F420mHdyLn0JIDzs6Fd4xKsMoacYkI6XAZ0arYqtyHn_yu5fiyQWH0lLfKZSoI7RZE0dIpb1YwdussP3kL-oIwFBDIoIt0cwUar_CviCKhnKzTsZcRY3OrpV6d0pDIm_IDUJQMo-WpA_a-Tr_5u4MrjU0aYOB4xNHEY8zsCT1hQibnPICUdmjee-yEDs2lHKhsrx1rNgHnGY0kvVXthRHrMRKaVnT3TP2Yv0vFev0zj_2kqkgwDLLmeZF24Thjo3rTFA4IdztQnb1lB4lCBLt7Mouzkg0u_jhicNFH99AoefCAQiWEfbDEVyM-eaXdR3YPDOj76f2ODsVaKsK4GNPj8KKj2v5o3yfO_bM7LBlvdMuLf_lAQsDa65VHBHVXy3sj31FT4_8v2dSHJKtOic-6fL2qlh_evkcgfCYxiDphxRgOMH2_YzTx92N1B3-6gob3tlej5Wt6Pgi21sYTALfiDHXkJ_Ax2uEQaEkSZODLWSS59TOOssmTidgMgy-X8AMqV3jncAPWPwMtoNJ0BwDUqjJV2AtVB1_zX2TfycBE0y0fiIDzrYMm1PF8eYbz-DLIvacmaaK9l6P-4hT8dzIQ-huZE5ZLcFU99OSG6SpE5b9FYf7zNZoB6AE0pN3WBTo9sFFIfvhYAVgg5Zy4B3IpOtbuq6x9hM4_01tMw**a029ddc2e2468f149929515340fb67e533cf8a4c27574fbb12fa34a7fbacba8e*AGxoiZ_0mO0ghuLI0hcnrDVOz-AMb-QAQ_zB5Pktj_M","_ga":"GA1.2.401110148.1591930422"}`

I want to access **a0:session** in above cookies in decrypt form. I am not sure if this is a bug or require different way of handling.

Thanks for help!

## To Reproduce

1. To reproduce this open any next.js project in gitpod.io for development
2. Create api in **pages/api/test** and add below code snippet
`export default async function test(req, res) {
        console.log(JSON.stringify(req.cookies));
}`
3. Above console.log will print multiple cookies one from gitpod and one from application

## Expected behavior

If multiple cookies are present in **req.cookies**, right cookie should be present in decrypt form or selecting right cookie decryt option should be given.  

## System information

- OS: Gitpod using Linux
- Browser: Chrome
- Version of Next.js: 9.4.4
- Version of Node.js: 12.16.3

## Additional context

This problem only present when multiple cookies are present in header request.

**Comment by jamesmosier** [comment, marked as answer]
Hi @jagyas! `req.cookies` will always contain all cookies passed to that request from the client. There is no way for `req.cookies` to know which cookie in particular you would want to retrieve.

**Reply by jagyas** [reply]
Thanks for your reply @jamesmosier

Do you have some idea, what will be the right way to handle this situation? 

I was thinking to use **js-cookie** or **nookies** npm packages to parse cookie after auth0.getSession() and then set in my global variable.

**Reply by jamesmosier** [reply]
Either of those packages would work well to parse the cookies and then you can get the cookie easily afterwards.

**Reply by jagyas** [reply]
I also figured out that there is no issue with Next.js. Actually somehow Auth0 is not appending token in cookies, on dev environments. It is something related with Apollo client which I am using. I found this in below repository:
https://github.com/vgrafe/nextjs-auth0-hasura

**Reply by Teelon** [reply]
I know this is an old thread but if anyone is interested. cookies-next is another package you can use.
https://www.npmjs.com/package/cookies-next

**Comment by surferwat** [comment]
One possible approach would be to 1) cleanup the _\_ with regex and then 2) parse it to construct an object. Thereafter, you should have access to _a0:session_.
